src/middleware/compression.rs

```diff
@@ -0,0 +1,420 @@
+                res.headers_mut()
+                    .append(CONTENT_ENCODING, encoding.to_header_value());
+                res
+            }
+            Encoding::Brotli => {
+                let mut body = Body::empty();
+                std::mem::swap(&mut body, res.body_mut());
+                let stream = stream::BrotliEncoder::new(body, self.settings.brotli);
+                *res.body_mut() = Body::from_stream(stream);
+                res.headers_mut()
+                    .append(CONTENT_ENCODING, encoding.to_header_value());
+                res
+            }
+            Encoding::Zstd => {
+                let mut body = Body::empty();
+                std::mem::swap(&mut body, res.body_mut());
+                let stream = stream::ZstdEncoder::new(body, self.settings.zstd);
+                *res.body_mut() = Body::from_stream(stream);
+                res.headers_mut()
+                    .append(CONTENT_ENCODING, encoding.to_header_value());
+                res
+            }
+            Encoding::Identity => res,
+        }
+    }
+}
+
+impl<Data: Send + Sync + 'static> Middleware<Data> for Compression {
+    fn handle<'a>(&'a self, cx: Context<Data>, next: Next<'a, Data>) -> BoxFuture<'a, Response> {
+        box_async! {
+            let encoding = match self.preferred_encoding(cx.headers()) {
+                Ok(encoding) => encoding,
+                Err(e) => return e.into_response(),
+            };
+            let res = next.run(cx).await;
+            self.encode(res, encoding)
+        }
+    }
+}
+
+/// Middleware for handling incoming request decompression.
+///
+/// This middleware currently supports HTTP decompression under the `gzip`, `deflate`, `br`, and `zstd` algorithms.
+#[derive(Debug, Default)]
+pub struct Decompression {}
+
+impl Decompression {
+    /// Creates a new Decompression middleware.
+    pub fn new() -> Self {
+        Self {}
+    }
+
+    fn parse_encoding(s: &str) -> Result<Encoding, Error> {
+        match s {
+            "gzip" => Ok(Encoding::Gzip),
+            "deflate" => Ok(Encoding::Deflate),
+            "br" => Ok(Encoding::Brotli),
+            "zstd" => Ok(Encoding::Zstd),
+            "identity" => Ok(Encoding::Identity),
+            _ => Err(Error::from(StatusCode::UNSUPPORTED_MEDIA_TYPE)),
+        }
+    }
+
+    fn decode(&self, req: &mut Request) -> Result<(), Error> {
+        let encodings = if let Some(hval) = req.headers().get(CONTENT_ENCODING) {
+            let hval = match hval.to_str() {
+                Ok(hval) => hval,
+                Err(e) => return Err(Error::from(StatusCode::BAD_REQUEST)),
+            };
+            hval.split(',')
+                .map(str::trim)
+                .rev() // apply decodings in reverse order
+                .map(Decompression::parse_encoding)
+                .collect::<Result<Vec<Encoding>, Error>>()?
+        } else {
+            return Ok(());
+        };
+
+        for encoding in encodings {
```

Comment:
Should we move this block to https://github.com/rustasync/tide/pull/194/files#diff-4069e2d3cfd22c835d307b45346dc68eR177 so we do not need to `collect` the encodings as vector and directly decompress after https://github.com/rustasync/tide/pull/194/files#diff-4069e2d3cfd22c835d307b45346dc68eR176 since the error rate there would be a lot lower.

Reply:
I mean I suppose we *could*, but the reason I do this here is because I don't think it should do unnecessary work unless it has to -- it helps if we know we can decode the request body from the getgo given a valid `Content-Encoding` header, and the overhead should usually be really cheap to begin with.

Say, for example, a client sends over a header `Content-Encoding: exi, gzip`, it's technically valid, *but* we don't know how to decode `exi`. If we moved this block there, we'd already be performing unnecessary work on a body we can't fully decode -- that is, per the reversal, we'd have `gzip, exi`, and after finishing decoding gzip we'd run into exi and then suddenly have to return the 415 error that `Decompression::parse_encoding` returns, `Err(Error::from(StatusCode::UNSUPPORTED_MEDIA_TYPE)`.